libraries/AP_Compass/AP_Compass_UAVCAN.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "AP_Compass_UAVCAN.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <drivers/drv_device.h>
+#include <drivers/drv_mag.h>
+#include <drivers/drv_hrt.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_mag_uavcan(level, fmt, args...) do { if ((level) <= AP_BoardConfig::get_can_debug()) { hal.console->printf(fmt, ##args); }} while (0)
+
+// There is limitation to use only one UAVCAN magnetometer now.
+
+/*
+  constructor - registers instance at top Compass driver
+ */
+AP_Compass_UAVCAN::AP_Compass_UAVCAN(Compass &compass):
+    AP_Compass_Backend(compass)
+{
+    if (hal.can_mgr != nullptr) {
+        AP_UAVCAN *ap_uavcan = hal.can_mgr->get_UAVCAN();
+        if (ap_uavcan != nullptr) {
+            // Give time to receive some packets from CAN if baro sensor is present
+            // This way it will get calibrated correctly
+            _instance = register_compass();
+            hal.scheduler->delay(2000);
+            ap_uavcan->register_mag_listener(this, 1);
+
+            struct DeviceStructure {
+                uint8_t bus_type : 3;
+                uint8_t bus: 5;
+                uint8_t address;
+                uint8_t devtype;
+            };
+            union DeviceId {
+                struct DeviceStructure devid_s;
+                uint32_t devid;
+            };
+            union DeviceId d;
+
+            d.devid_s.bus_type = 3;
+            d.devid_s.bus = 0;
+            d.devid_s.address = 0;
+            d.devid_s.devtype = 0;
```

Comment:
Yes, it was talked, but actually you do not have any IDs for other magnetometers as well. Only type is differentiated.